Optimize job pickling to speed up Manager thread performance on large graphs

RaphaelLilt opened the pull request:
**Description**
We found that the manager threads were starting new jobs deep in a large graph slowly (approximately one job every 5 seconds on average). When profiling with `py-spy`, we found that the pickling in `sisyphus.job.Job._sis_setup_directory` was a performance bottleneck. Additionally, the pickled jobs files (`job.save`) were very large—reaching up to 104 MB for jobs that only had file paths as input—which is as we found out since they contain all their preprocessor jobs (e.g. `job.input_files[0].creator.input_files[0].creator...`).

**Proposed Changes**
- Modify `sisyphus.job_path.AbstractPath.__getstate__` used by Pickle to set the `creator` to `None` and adding a method `__sis_state__` used by `sisyphus.hash.get_object_state`. The worker running a job only needs to know the jobs input paths but not their creators.
- This ensures that the hashes don't change and the pickled jobs only contain the necessary state.
- In `global_constants.py` add `INCLUDE_CREATOR_STATE` to control the inclusion of `"creator"` fields in serialized `AbstractPath`.

**Results:**
- **Increased Speed:** After the change we were able to simultaneously run 10,000 jobs compared to approximately 5 Jobs before for large graphs.
- **Reduced Pickled File Size**: The pickled job files are much smaller about 100 KB compared to the previous > 104 MB.

This optimization leads to much faster job initializations for large graphs making the system to scale to larger graphs.

critias commented:
Why not call it `__sis_state__` instead of `_get_state_with_creator` and make use of the already existing check? Thinking about that, `__sis_state__` should have been called something like `__sis_hash_state__`...

It use to be implemented in a similar way before we switched to pickle. At the time the pickled files were only a few MB, so we kept everything inside the `job.save`. That way you can just unpickle the job and look throw the whole tree if you wanted to debug something, but with 100MB files it's not worth it.

I would still recommend to make it an option so it can be turned back on in case it's needed for debugging.

michelwi commented on sisyphus/hash.py:
```diff
@@ -4,6 +4,8 @@
 import pathlib
 from inspect import isclass, isfunction, ismemberdescriptor
 
+from sisyphus.job_path import AbstractPath
```
Thank you for taking the time of profiling this and additionally for taking the time to contribute the changes to our repo!

I noticed that our tests are failing. `black` can be fixed trivially later, but I think the CI/test might indicate a problem.
We have a .. unorthodox way of importing the `global_settings` so that they can be overridden by user settings in a separate file. This leads to problems with circular imports: in `hash` we now added an import from `job_path` which in turn imports `global_settings` which imports `hash`. And the symptoms can be found in the CI pipeline.
Maybe @critias or @JackTemaki can explain the problem better 🙈 

Maybe moving the import into the `get_object_state` function would already solve it by breaking the loop?

critias commented:
You are right, that's a circular import. Your suggestion of moving the import should work, but I would prefer the change I suggested above to just renaming `_get_state_with_creator ` to `__sis_state__` and revert all changes in `sisyphus/hash.py`.

RaphaelLilt commented:
Thanks @critias and @michelwi, I'll incorporate your suggestions and fix the CI tests.

michelwi commented on sisyphus/global_settings.py:
```diff
@@ -205,6 +205,9 @@ def file_caching(path):
 SHOW_JOB_TARGETS = True
 
+# Controls whether to include 'creator' in serialized state of an abstract path
+INCLUDE_CREATOR_STATE = True
```
I think we could even default to False and leave the option to be enabled for debugging..?

critias replied:
I agree, I don't see how enabling it by default would cause any problems and if it does we can simply set it to True again.

RaphaelLilt replied:
Makes sense, I changed it.

RaphaelLilt commented:
@michelwi I fixed the unit test. For that I removed the `deepcopy` in `sisyphus.job.JobSingleton` which seems to be unnecessary for me and would give an additional speedup. It would be great if you could have a look and verify that this changes does not break anything.

michelwi commented:
There is some issue with the hashes now. In one example that I looked into there is a `i6_core.returnn.compile.CompileTFGraphJob`  that takes as input a `i6_core.returnn.config.ReturnnConfig` object. In the config I define the dataset as a nested dict that contains `tk.Path` objects which are outputs of `i6_core.returnn.hdf.ReturnnDumpHDFJob`s.
(all referring to https://github.com/rwth-i6/i6_core)

In the current master branch, I have (output of the sis console)
```
In [1]: js = tk.find_job("Compile")
Jobs found:
0: Job<work/i6_core/returnn/compile/CompileTFGraphJob.yK7hwaw8zNIC>

In [2]: p = js[0].__dict__["returnn_config"].config["train"]["files"][0]

In [3]: print(f"{p.creator=} {p.path=}")
p.creator=Job<work/i6_core/returnn/hdf/ReturnnDumpHDFJob.6SGIal2fRE84> p.path='data.hdf'

In [4]: j = tk.find_job("ReturnnDumpHDFJob.6SGIal2fRE84")[0] 
Jobs found:
0: Job<work/i6_core/returnn/hdf/ReturnnDumpHDFJob.6SGIal2fRE84>

In [5]: j.out_hdf.creator
Out[5]: Job<work/i6_core/returnn/hdf/ReturnnDumpHDFJob.6SGIal2fRE84>
```
but with this update there is
```
In [1]: js = tk.find_job("Compile")
Jobs found:
0: Job<work/i6_core/returnn/compile/CompileTFGraphJob.7mZhYhttecOr>

In [2]: p = js[0].__dict__["returnn_config"].config["train"]["files"][0]

In [3]: print(f"{p.creator=} {p.path=}")
p.creator=None p.path='/home/wmichel/code/work/i6_core/returnn/hdf/ReturnnDumpHDFJob.6SGIal2fRE84/output/data.hdf'

In [4]: j = tk.find_job("ReturnnDumpHDFJob.6SGIal2fRE84")[0] 
Jobs found:
0: Job<work/i6_core/returnn/hdf/ReturnnDumpHDFJob.6SGIal2fRE84>

In [5]: j.out_hdf.creator
Out[5]: Job<work/i6_core/returnn/hdf/ReturnnDumpHDFJob.6SGIal2fRE84>
```

Here the hash of the `CompileTFGraphJob` changed and the path that it holds has its creator removed. But the output of the `ReturnnDumpHDFJob` has the creator set correctly in both cases.

michelwi commented:
The issue seems to be that we like to do deepcopies of our config objects if they are given to multiple jobs and that deepcopy uses `__getstate__` to create the copy, which now breaks the dependence between the jobs.

So, @critias, I am afraid we cannot use your elegant solution, or we need to additionally implement a `__deepcopy__` method that uses `__sis_state__` instead of `__getstate__`.

michelwi reviewed (changes requested):
request changes as current state breaks existing setups

michelwi commented:
Btw. @RaphaelLilt, I assume you are also using sisyphus as your workflow manager. So if you also want to be included in design decisions or checks if changes break existing setups, then feel free to raise a hand or even submit some test cases.

RaphaelLilt commented:
Thanks, @michelwi, for the comments. Could you please share a minimal example showing that the hashes change between the two versions? I can't reproduce that.
What's the reasoning behind passing a deepcopy of the arguments to the `_sis_init` method? Shouldn't that result in performance issues, similar to the problems that arise when picking up jobs for large graphs this PR intends to fix?

michelwi commented:
> please share a minimal example

certainly.

```
import copy
from sisyphus import Job, tk

class WriteJob(Job):
    def __init__(self, some_val):
        self.some_val = some_val
        self.out = self.output_path("out.txt")

    def run():
        with open(self.out, 'wt') as outpath:
            outpath.write(self.some_val)

class ReadJob(Job):
    def __init__(self, f):
        self.f = f
        self.out = self.output_path("out.txt")

    def run():
        with open(self.f) as infie:
            for line in infile:
                print(line)
        with open(self.out, "wt") as outfile:
            outfile.write("done")


def py():
    f = WriteJob("Hello World").out
    o = ReadJob(f).out
    tk.register_output("output", o)
    o2 = ReadJob(copy.deepcopy(f)).out
    tk.register_output("output2", o2)
```

expected behavior:
```
$ sis c config/test_config.py

In [1]: tk.find_job("Job")
Jobs found:
0: Job<work/config/test_config/WriteJob.mlrz8RnG9o9U>
1: Job<work/config/test_config/ReadJob.PrhZGzhQOgKP>

$ sis m config/test_config.py
[...]
[2024-10-25 14:53:45,063] INFO: Experiment directory: /home/wmichel/code      Call: /home/wmichel/code/sisyphus/sis m config/test_config.py
[2024-10-25 14:53:45,063] INFO: runnable: Job<work/config/test_config/WriteJob.mlrz8RnG9o9U> <target: output>
[2024-10-25 14:53:45,063] INFO: runnable(1) waiting(1)
```

observed behavior
```
$ sis c config/test_config.py

In [1]: tk.find_job("Job")
Jobs found:
0: Job<work/config/test_config/ReadJob.PrhZGzhQOgKP>
1: Job<work/config/test_config/WriteJob.mlrz8RnG9o9U>
2: Job<work/config/test_config/ReadJob.YSB3u3nSRZJN>

$ sis m config/test_config.py
[...]
[2024-10-25 14:53:28,383] INFO: Experiment directory: /home/wmichel/code      Call: /home/wmichel/code/sis_lillt/sis m config/test_config.py
[2024-10-25 14:53:28,383] ERROR: input_missing: /home/wmichel/code/work/config/test_config/WriteJob.mlrz8RnG9o9U/output/out.txt
[2024-10-25 14:53:28,384] INFO: runnable: Job<work/config/test_config/WriteJob.mlrz8RnG9o9U> <target: output>
[2024-10-25 14:53:28,384] INFO: input_missing(1) runnable(1) waiting(2)
```

The hash of the Job without deepcopy is the same, but in the second case there is one additional Job with a different hash.
Also the manager complains about a missing input, because the deepcopy breaks the dependence between the jobs and sisyphus now assumes that `WriteJob.mlrz8RnG9o9U/output/out.txt` must be an external input.


> What's the reasoning behind passing a deepcopy of the arguments to the _sis_init method?

I was not referring to this particular deepcopy, but rather to generally using deepcopy in my recipes. Imagine that I have prepared my NN training config object (e.g. some very deep nested dict) and now want to do a parameter sweep in training, e.g.
```
config_template = get_it_somehow()
for num_layers in [10, 12, 14, 16]:
    my_config = copy.deepcopy(config_template)
    my_config["params"]["network_config"]["num_layers"] = num_layers
    j = TrainingJob(my_config)
```
then I would employ deepcopy to multiply my config and to avoid side effect interactions between my experiments.

RaphaelLilt commented:
@michelwi I just pushed a change that should fix your deepcopy issues. Could you please have a look?

michelwi commented:
I was just copy pasting the next error message:

```
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/local/lib/python3.10/copy.py", line 153, in deepcopy
    y = copier(memo)
  File "/opt/atlassian/pipelines/agent/build/.nox/check_hashes-3-10/lib/python3.10/site-packages/sisyphus/job_path.py", line 252, in __deepcopy__
    return self.__class__(**state)
TypeError: Variable.__init__() got an unexpected keyword argument 'cached'
```
But I saw you updated again - will test that now.

critias commented:
LGTM, but I leave the final approval to @michelwi since he has been testing it.

michelwi reviewed (approved):
I am not so deep into python to know if the deepcopying is done correctly, but all tests pass and an LLM thought the code was good.

RaphaelLilt commented:
Thanks @critias and @michelwi for the reviews. How do we merge the PR to master?

albertz commented:
Just as a comment, I wonder if this change could be problematic for other cases. Now here we have the assumption that whenever some Path is pickled/deepcopied, we never want to know about the creator. This might be true for the example given here (and even fix the issues of having too large job.save files). But this assumption is probably not true in general. In general, when you pickle/deepcopy something, you would expect that it contains everything.

It could maybe be solved by having some context manager or so which controls the pickle/deepcopy logic, so when we pickle for job.save, we enable the context manager, which removes some parts for the copy, but in general, when it is not enabled, it would follow the normal expected behavior that everything is copied.

But maybe we can just leave it like it is now until we really get to this problem. However, it might be hard to find the problem later if sth goes wrong because of this.

michelwi commented:
> whenever some Path is pickled/deepcopied

Your comment is valid for pickling Jobs.
For deepcopy we indeed already ran into a problem and have a separate implementation now that keeps the creator.

RaphaelLilt commented:
> Your comment is valid for pickling Jobs.

That's right.

>  In general, when you pickle/deepcopy something, you would expect that it contains everything.

I think this flexibility comes with a huge performance cost which makes running larger graphs with >1M jobs extremely slow. We currently trying to make `remove_job_and_descendants` faster, since it takes multiple days.

albertz commented:
Yes, I'm not against the solution here, I was just saying, this solution is now specific to this problem (and a very reasonable solution for it), but maybe it breaks now in other cases, as this behavior now is a bit unexpected. But I don't think we need to change anything for now.

In general, it's great that you tackle this problem on working with large graphs. I was often frustrated by the speed, and also have done some other optimizations already in the past.

patrick-wilken commented:
@Zettelkasten noticed these changes break the `TASK_INPUTS_MUST_BE_AVAILABLE = False` case, which was the default/only behavior until recently.

The worker calls `Path.available()` which has two branches depending on whether the Path has a creator. If yes, it is just checked whether the the creator job has finished. If not, the existence of the file is checked, which does not work for the cases where `TASK_INPUTS_MUST_BE_AVAILABLE` matters (mostly path prefixes).

I don't like that we use these prefixes in our recipes, I think they should have never been supported, and we will get rid of them in the long run. But it shows that what @albertz says is correct, by default you would expect to have the complete Path object available in the worker and not having that is error prone.

patrick-wilken commented:
When I said "supported" I assumed `Job.path_available()`was implemented with these prefixes in mind. But @michelwi said checking whether the Job is finished is probably a shortcut to avoid file system accesses. This would make sense, and in this case you have every right to break path prefixes that happened to work just because of that. 😄

RaphaelLilt commented:
I'm not sure if I can follow. The only place where `TASK_INPUTS_MUST_BE_AVAILABLE` appears in the repo is [here](https://github.com/rwth-i6/sisyphus/blob/d256f43b69c0dd668b52822ec0016cb8479e5efd/sisyphus/task.py#L153-L158). I don’t see where the worker calling `Path.available()` and whether the `Path` has a creator. From what I see, it’s just checking the file system directly. Could you clarify?

patrick-wilken commented:
[This](https://github.com/rwth-i6/sisyphus/blob/d256f43b69c0dd668b52822ec0016cb8479e5efd/sisyphus/worker.py#L213) `job._sis_runnable()` check leads to a `Path.available()` call in `Job._sis_all_path_available()`.

So it's not directly in the `TASK_INPUTS_MUST_BE_AVAILABLE` related code, but it's logically related. If `TASK_INPUTS_MUST_BE_AVAILABLE` is false, paths are allowed to be file name prefixes and that's exactly the case where the result of `Path.available()` changes depending on whether the creator is kept or removed, because the check changes from "creator job is finished" to "file exists" and only then those can be different.

(`TASK_INPUTS_MUST_EXIST` would be a more exact name.)
